### PR TITLE
lemmas on `subseq` and `rot`

### PR DESCRIPTION
**Pull request by chdoc**
##### Motivation for this change

Lemmas for reasoning about `subseq`, in particular subsequences of duplicate free sequences and rotations of sequences. These turned out to be useful when formalizing subcycles (i.e., subsequences up to rotation).

@CohenCyril I chose `_pivot` for the `s1 ++ x :: s2` pattern, but I am open to suggestions. I would favor this pattern over `rcons s1 x ++ s2`, because the former arises naturally when using `rot_to_arc` (`x` being the beginning of the second arc). 

##### Things done/to do

- [x] added corresponding entries in `CHANGELOG_UNRELEASED.md` (do not edit former entries)
- [ ] ~added corresponding documentation in the headers~

##### Automatic note to reviewers

Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-following,-reviewing-and-playing-with-a-PR#checklist-for-reviewing-a-pr) and make sure there is a milestone.

### CODE REVIEW

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [inline mathcomp/ssreflect/seq.v:1398]: `x \notin s3 ++ s4` looks sufficient

Comment (chdoc) [reply mathcomp/ssreflect/seq.v:1398]: I guess you're right, the statement was motivated by what's available at the point of use in `subseq_pivot`. There, I suppose, the same criticism applies. I will have a look later; the proof of `subseq_pivot` is based on `subseq_uniqP`...

**Comment by chdoc** [comment]
Well, just ran a count on my graph-theory development and I have about the same number of uses of `path.splitP` (if one imports `all_ssreflect`, `splitP` ends up being the one for ordinals) and of `splitPr`. The former is more useful when dealing with paths, where one usually cares about the last element, so `splitP` allows rewriting with `last_rcons`. The latter is more useful when dealing with arcs, where the "end" of the arc is actually part of the other arc (i.e, `arc s x y` contains `x` but not `y`).

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [reply mathcomp/ssreflect/seq.v:1398]: On second thought, I'm not sure what the ideal statement for this lemma is. The proposed `x \notin x3 ++ s4` is clearly sufficient. But then one has to go via `count (pred1 x)` to recover `x \notin s1` to prove the lemma. In fact, that would be the only purpose of assuming `x \notin s4`. So one could weaken the assumption even further to assume `x \notin s1` and `x \notin s3`. On the other hand, this would make the lemma even more cumersome to use if one only has `uniq (s3 ++ x :: s4)` in the context. In my setting, almost all occurring sequences are duplicate free.  :thinking:

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [inline mathcomp/ssreflect/seq.v:1934]:
```suggestion
Lemma rev_mask m s : size m = size s -> rev (mask m s) = mask (rev m) (rev s).
Proof.
move: m s; apply: seq_ind2 => //= b x m s eq_size_sm IH.
by case: b; rewrite !rev_cons mask_rcons ?IH ?size_rev// (cats1, cats0).
Qed.
```
provided you also add
```coq
Lemma mask_rcons b m x s : size m = size s ->
  mask (rcons m b) (rcons s x) = mask m s ++ nseq b x.
Proof. by move=> ms; rewrite -!cats1 mask_cat//; case: b. Qed.
```

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [inline mathcomp/ssreflect/seq.v:2310]: I suggest `subseq_uniq_pivot` (using the same convention as `eqseq_uniq_pivotr`, but omitting the `r` because there would never be a `l` version)

**Review by CohenCyril — CHANGES_REQUESTED** [review]
I think this PR is good to merge, modulo some minor changes.

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [inline mathcomp/ssreflect/seq.v:1398]:
```suggestion
Lemma eqseq_uniq_pivotr s1 s2 s3 s4 x : uniq (s3 ++ x :: s4) ->
  s1 ++ x :: s2 == s3 ++ x :: s4 = (s1 == s3) && (s2 == s4).
```

Comment (chdoc) [reply mathcomp/ssreflect/seq.v:1398]: Hmm, according to `mainSymbol_suffixes(_condition)?` I would rather name this `eqseq_pivot_uniqr`. The "pivot" is `x` and the `pivot` is to describe the `_ ++ _ :: _` pattern. (I guess neither of us wants to call it `uniqr_eqseq_pivot` after the `(condition_)?mainSymbol_suffixes` convention.). I guess a `uniql` variant doesn't hurt either.

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [reply mathcomp/ssreflect/seq.v:2310]: Hmm, if I follow my argument from above, I end up at `subseq_pivot_uniq` which is ambiguous as uniq is unary predicate. So maybe `uniq_subseq_pivot` is more appropriate. Maybe naming the other one `uniqr_eqseq_pivot` isn't that bad a choice after all. @CohenCyril What do you think?

Comment (chdoc) [reply mathcomp/ssreflect/seq.v:2310]: (Hmm, there was no "Reply..." for your first comment, so I commented after clicking on "view changes", but this comment isn't shown above :thinking: )

**Comment by chdoc** [comment]
@CohenCyril , I rebased, but I kept the changes following your review in a separate commit. Please ping me before merging, even if you are okay with my alternative choice of names, so that I can squash the commits.

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [inline mathcomp/ssreflect/seq.v:1420]: While we are at it, we could include all the following variations and side lemmas:
(and for the sake of uniformity, with `uniq_subseq_pivot` I would prefer the names `uniq_eqseq_pivotl` and `uniq_eqseq_pivotr`)
```suggestion
Lemma index_pivot x s1 s2 (s := s1 ++ x :: s2) : x \notin s1 -> 
  index x s = size s1.
Proof. by rewrite index_cat/= eqxx addn0; case: ifPn. Qed.

Lemma take_pivot x s2 s1 (s := s1 ++ x :: s2) : x \notin s1 -> 
  take (index x s) s = s1.
Proof. by move=> /index_pivot->; rewrite take_size_cat. Qed.

Lemma rev_pivot x s1 s2 : rev (s1 ++ x :: s2) = rev s2 ++ x :: rev s1.
Proof. by rewrite rev_cat rev_cons cat_rcons. Qed.

Lemma eqseq_pivot2l x s1 s2 s3 s4 : x \notin s1 -> x \notin s3 ->
  s1 ++ x :: s2 == s3 ++ x :: s4 = (s1 == s3) && (s2 == s4).
Proof.
move=> xNs1 xNs3; apply/idP/idP => [E|/andP[/eqP-> /eqP->]//].
suff S : size s1 = size s3 by rewrite eqseq_cat// eqseq_cons eqxx in E.
by rewrite -(index_pivot s2 xNs1) (eqP E) index_pivot.
Qed.

Lemma eqseq_pivot2r x s1 s2 s3 s4 : x \notin s2 -> x \notin s4 ->
  s1 ++ x :: s2 == s3 ++ x :: s4 = (s1 == s3) && (s2 == s4).
Proof.
move=> xNs2 xNs4; rewrite -(can_eq revK) !rev_pivot.
by rewrite eqseq_pivot2l ?mem_rev// !(can_eq revK) andbC.
Qed.

Lemma eqseq_pivotl x s1 s2 s3 s4 : x \notin s1 -> x \notin s2 ->
  s1 ++ x :: s2 == s3 ++ x :: s4 = (s1 == s3) && (s2 == s4).
Proof.
move=> xNs1 xNs2; apply/idP/idP => [eqs|/andP[/eqP-> /eqP->]//].
rewrite -(@eqseq_pivot2l x)//; have /eqP/(congr1 (count_mem x)) := eqs.
rewrite !count_cat/= eqxx !addnS (count_memPn _ _ xNs1) (count_memPn _ _ xNs2).
by move=> -[/esym/eqP]; rewrite addn_eq0 => /andP[/eqP/count_memPn].
Qed.

Lemma eqseq_pivotr x s1 s2 s3 s4 : x \notin s3 -> x \notin s4 ->
  s1 ++ x :: s2 == s3 ++ x :: s4 = (s1 == s3) && (s2 == s4).
Proof.
by move=> *; rewrite eq_sym eqseq_pivotl// [s3 == _]eq_sym [s4 == _]eq_sym.
Qed.

Lemma uniq_eqseq_pivotl x s1 s2 s3 s4 : uniq (s1 ++ x :: s2) ->
  s1 ++ x :: s2 == s3 ++ x :: s4 = (s1 == s3) && (s2 == s4).
Proof.
by rewrite uniq_catC/= mem_cat => /andP[/norP[? ?] _]; rewrite eqseq_pivotl.
Qed.

Lemma uniq_eqseq_pivotr x s1 s2 s3 s4 : uniq (s3 ++ x :: s4) ->
  s1 ++ x :: s2 == s3 ++ x :: s4 = (s1 == s3) && (s2 == s4).
Proof.
by move=> ?; rewrite eq_sym uniq_eqseq_pivotl// [s3 == _]eq_sym [s4 == _]eq_sym.
Qed.
```

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [inline mathcomp/ssreflect/seq.v:2147]: shorter proof:
```suggestion
by rewrite (leq_trans (count_size _ _))// size_take; case: ltnP.
```

**Comment by CohenCyril** [comment]
I guess it's time to summon @pi8027...

File: mathcomp/ssreflect/seq.v
Comment (pi8027) [inline mathcomp/ssreflect/seq.v:2159]:
```suggestion
by case/subseqP => m size_m ->; rewrite rev_mask // mask_subseq.
```

File: mathcomp/ssreflect/seq.v
Comment (pi8027) [inline mathcomp/ssreflect/seq.v:2173]:
```suggestion
case/subseqP=> m size_m ->.
```

File: mathcomp/ssreflect/seq.v
Comment (pi8027) [inline mathcomp/ssreflect/seq.v:2163]: There is an `Implicit Types` declaration for `s` in this section.
```suggestion
Lemma subseq_cat2l s s1 s2 : subseq (s ++ s1) (s ++ s2) = subseq s1 s2.
```

File: mathcomp/ssreflect/seq.v
Comment (pi8027) [inline mathcomp/ssreflect/seq.v:2167]:
```suggestion
Lemma subseq_cat2r s s1 s2 : subseq (s1 ++ s) (s2 ++ s) = subseq s1 s2.
```

File: mathcomp/ssreflect/seq.v
Comment (pi8027) [inline mathcomp/ssreflect/seq.v:2378]: The conclusion can be equivalence rather than implication.
```suggestion
Lemma uniq_subseq_pivot s1 s2 s3 s4 x (s := s3 ++ x :: s4) :
  uniq s -> subseq (s1 ++ x :: s2) s = (subseq s1 s3 && subseq s2 s4).
Proof.
move=> uniq_s; apply/idP/idP; last first.
  case/andP=> /subseqP [m1 size_m1 ->] /subseqP [m2 size_m2 ->].
  by rewrite -(mask_cat (true :: _) (_ :: _)) // mask_subseq.
move=> sub_s'_s; have uniq_s' := subseq_uniq sub_s'_s uniq_s.
have/eqP {sub_s'_s uniq_s} := subseq_uniqP _ uniq_s sub_s'_s.
rewrite filter_cat /= mem_cat mem_head orbT /= uniq_eqseq_pivotl //.
by case/andP=> /eqP -> /eqP ->; rewrite !filter_subseq.
Qed.
```

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [reply mathcomp/ssreflect/seq.v:2173]: I prefer the former.

Comment (CohenCyril) [reply mathcomp/ssreflect/seq.v:2173]: except for the extra spaces
```suggestion
move=> /subseqP[m size_m ->].
```

File: CHANGELOG_UNRELEASED.md
Comment (CohenCyril) [inline CHANGELOG_UNRELEASED.md:256]:
```suggestion
  `uniq_eqseq_pivotl`, `uniq_eqseq_pivotr`, `mask_rcons`, `rev_mask`,
```

File: mathcomp/ssreflect/seq.v
Comment (pi8027) [reply mathcomp/ssreflect/seq.v:2173]: > I prefer the former.

Why do you think so?

File: mathcomp/ssreflect/seq.v
Comment (pi8027) [reply mathcomp/ssreflect/seq.v:2159]: @CohenCyril You may prefer `move=> /subseqP[m size_m ->]` over my suggestion here too.

File: mathcomp/ssreflect/seq.v
Comment (chdoc) [reply mathcomp/ssreflect/seq.v:2378]: Of course, this could be an equality, but I don't think the direction from left to right is of much use. In particular, the direction from right to left can (after intros) be simplified to `by rewrite cat_subseq //= eqxx.` The main use of this lemma is to split the complicated assumption into simpler (more useful) ones.

Comment (pi8027) [reply mathcomp/ssreflect/seq.v:2371]: If you want to use it mainly as a view lemma, stating it using an equation can lengthen proof scripts a bit indeed. On the other hand, if you have only the implication version of this lemma and want to rewrite a subterm of the goal or hypotheses with it, I think it will be even costly.

Comment (pi8027) [reply mathcomp/ssreflect/seq.v:2371]: At least, I prefer to use `&&` for conjunctions of boolean values.

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [reply mathcomp/ssreflect/seq.v:2166]: The syntax `move/view` and `case/view` is the historical syntax, views in intro pattern came later AFAIR , and for the sake of teaching, we usually only teach the latter, so whenever it does not save a line, I personally prefer when the view, case and dispatch are all regrouped in the intro pattern as in `/view[stuff]`, it looks more "modern" to me.

Comment (pi8027) [reply mathcomp/ssreflect/seq.v:2166]: @CohenCyril I didn't know that. Thanks!

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [reply mathcomp/ssreflect/seq.v:2371]: I'm not sure which one I like best (I think I prefer the equality, since it is more versatile, even though there is a slight overhead when one wants to transform the hypothesis, (hopefully it might be fixed by #501 someday)), but if one keeps the implication, I'd rather have a `prod` than a  `and` in the conclusion, so as to rewrite with it, at least. BTW the variable `x` should come first in case it needs to be specified.

Comment (chdoc) [reply mathcomp/ssreflect/seq.v]: Fine, you get your equality.

File: mathcomp/ssreflect/seq.v
Comment (CohenCyril) [inline mathcomp/ssreflect/seq.v:2383]: One last detail: `x` should come first.

Comment (CohenCyril) [reply mathcomp/ssreflect/seq.v:2383]:
```suggestion
Lemma uniq_subseq_pivot x (s1 s2 s3 s4 : seq T) (s := s3 ++ x :: s4) :
```

Comment (chdoc) [reply mathcomp/ssreflect/seq.v:2383]: I knew I forgot something ...

**Comment by chdoc** [comment]
And another rebase to resolve the conflicts with the other recent additions to `seq.v`...

**Review by CohenCyril — APPROVED** [review]
Ready to merge.

**Comment by CohenCyril** [comment]
> And another rebase to resolve the conflicts with the other recent additions to `seq.v`...

It might very well have been the last one though.

**Pull request merged**
